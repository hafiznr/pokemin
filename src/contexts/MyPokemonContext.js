import React, { createContext, useState } from 'react';

export const MyPokemonContext = createContext();

export const MyPokemonContextProvider = ({ children, initialData }) => {
  const [myPokemonData, setMyPokemonData] = useState(JSON.parse(initialData) || []);

  const addPokemon = (data) => {
    const newPokemonData = [...myPokemonData, data];

    setMyPokemonData(newPokemonData);

    localStorage.setItem('mypokemon', JSON.stringify(newPokemonData));
  }

  const removePokemon = (nickname) => {
    const newPokemonData = myPokemonData
      .filter(pokemon => pokemon.nickname !== nickname);

    setMyPokemonData(newPokemonData);

    localStorage.setItem('mypokemon', JSON.stringify(newPokemonData));
  }

  return (
    <MyPokemonContext.Provider 
      value={{ myPokemonData, addPokemon, removePokemon }}
    >
      {children}
    </MyPokemonContext.Provider>
  )
}
