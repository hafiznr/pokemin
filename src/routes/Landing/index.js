/** @jsxRuntime classic */
/** @jsx jsx */
import React, { useEffect, useState } from 'react';
import { css, jsx } from '@emotion/react';
import { useQuery, gql } from '@apollo/client';
import PokemonCard from '../../components/PokemonCard';
import { Box } from '@chakra-ui/layout';
import { useContext } from 'react';
import { MyPokemonContext } from '../../contexts/MyPokemonContext';
import TopBar from '../../components/TopBar';

const GET_POKEMONS = gql`
  query pokemons($limit: Int, $offset: Int) {
    pokemons(limit: $limit, offset: $offset) {
      count
      next
      previous
      status
      message
      results {
        id
        url
        name
        image
      }
    }
  }
`;

const Landing = () => {
  const { myPokemonData } = useContext(MyPokemonContext);
  
  const [offset, setOffset] = useState(0);
  const [pokemonList, setPokemonList] = useState([]);

  const ownedData = {};
  myPokemonData.forEach(item => {
    if (ownedData[item.id]) {
      ownedData[item.id]++;
    } else {
      ownedData[item.id] = 1;
    }
  });

  const handleSetPokemonList = (data) => {
    const { pokemons } = data || {};
    const { results } = pokemons || {};

    setPokemonList([...pokemonList, ...results]);
  }

  const { loading, error } = useQuery(GET_POKEMONS, {
    variables: {
      limit: 10,
      offset
    },
    onCompleted: handleSetPokemonList
  });

  useEffect(() => {
    const handleScroll = () => {
      if (window.innerHeight + document.documentElement.scrollTop 
        === document.documentElement.offsetHeight) {
        setOffset(offset + 10);
      }
    }

    window.addEventListener("scroll", handleScroll);

    return () => window.removeEventListener("scroll", handleScroll);
  }, [offset]);

  return (
    <React.Fragment>
      <TopBar />
      <Box padding="8px 16px 90px 16px">
        <div>
          {pokemonList.map(pokemon => {
            return (
              <PokemonCard
                id={pokemon.id}
                name={pokemon.name}
                image={pokemon.image}
                owned={ownedData[pokemon.id] || 0}
                key={pokemon.id}
              />
            )
          })}
        </div>
      </Box>
    </React.Fragment>
  );
}

export default Landing;
