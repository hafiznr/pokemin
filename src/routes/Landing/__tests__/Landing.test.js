import { ApolloProvider } from '@apollo/client';
import { render, screen } from '@testing-library/react';
import Landing from '../';
import { MyPokemonContextProvider } from '../../../contexts/MyPokemonContext';
import { client }  from '../../../client';

test('renders Landing', () => {
  const pokemonData = [
    {
      id: 4,
      name: 'charmander',
      image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png",
      nickname: 'ADsadasdas'
    },
    {
      id: 2,
      name: 'ivysaur',
      image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png",
      nickname: '0'
    },
    {
      id: 4,
      name: 'charmander',
      image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png",
      nickname: 'hamamama'
    },
  ];

  const { container } = render(
    <ApolloProvider client={client}>
      <MyPokemonContextProvider initialData={JSON.stringify(pokemonData)}>
        <Landing />
      </MyPokemonContextProvider>
    </ApolloProvider>
  );

  expect(container).toMatchSnapshot();
});