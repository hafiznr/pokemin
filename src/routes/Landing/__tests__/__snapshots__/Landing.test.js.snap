// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Landing 1`] = `
<div>
  <div
    class="css-1wxv3d5"
  >
    <div
      class="chakra-stack css-n21gh5"
    >
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
      <div
        class="chakra-skeleton css-1ylkj6c"
      />
    </div>
  </div>
</div>
`;
